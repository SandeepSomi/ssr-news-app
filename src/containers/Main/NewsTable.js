import React from 'react';
import { useSelector, useDispatch } from 'react-redux'
import moment from 'moment';
import * as actions from './actions';
import cn from 'classnames';

export default function () {
  const dispatch = useDispatch();
  const page = useSelector(state => state.main.page);
  const nbPages = useSelector(state => state.main.nbPages);
  const hitsPerPage = useSelector(state => state.main.hitsPerPage);
  const hits = useSelector(state => state.main.hits);

  function onHide(objectID) {
    dispatch(actions.hideNews(objectID));
  }

  function handlePageChange(nextPage) {
    if (nextPage < 0 || nextPage > nbPages) return null;
    dispatch(actions.search({ page: nextPage, hitsPerPage }))
  }

  function upvote(objectID) {
    dispatch(actions.upvote(objectID));
  }

  function rowRenderer(hit, index) {
    const { title, num_comments, author, created_at, objectID, url, votes = 0, isHidden = false } = hit
    if (isHidden) {
      return null
    }
    const origin = url && url.split('/')[2] || '';

    return (
      <tr key={objectID || index}>
        <td>{num_comments}</td>
        <td>{votes}</td>
        <td>
          <span onClick={() => upvote(objectID)}>
            <i className="fas fa-arrow-alt-circle-up"></i>
          </span>
        </td>
        <td>
          {title}
          {origin ? <span>
            <a href={url} target="_blank">({origin})</a>
          </span> : null}
          <span className="news-by">by</span>
          <span>{author}</span>
          <span className="posted">{moment(created_at).fromNow(true)}</span>
          <span className="hide-action" onClick={() => onHide(objectID)}>[ hide ]</span>
        </td>
      </tr>
    )
  }

  // if (isLoading) {
  //   return (
  //     <div className="vh-100 d-flex justify-content-center align-items-center">
  //       <div>
  //         <h6 className="text-secondary">Loading...</h6>
  //       </div>
  //     </div>
  //   )
  // }

  return (
    <div className="news-table-wrapper">
      <table id="news-table">
        <thead>
          <tr>
            <th>Comments</th>
            <th>Vote Count</th>
            <th>UpVote</th>
            <th>NewsDetail</th>
          </tr>
        </thead>
        <tbody>
          {hits.map(rowRenderer)}
        </tbody>
        <tfoot>
          <tr>
            <td colSpan="4">
              <span className="float-left">{page + 1}</span>
              <span className={cn({ 'disabled': page <= 0 })} onClick={() => handlePageChange(page - 1)}>Previous</span>
              {' | '}
              <span className={cn({ 'disabled': page >= nbPages })} onClick={() => handlePageChange(page + 1)}>Next</span>
            </td>
          </tr>
        </tfoot>
      </table>
    </div>
  )
}