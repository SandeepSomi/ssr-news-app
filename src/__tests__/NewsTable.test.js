import React from 'react';
import thunk from 'redux-thunk'
import configureMockStore from 'redux-mock-store'
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import jest from 'jest';
import NewsTable from '../containers/Main/NewsTable';

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

const store = mockStore({
    main: {
        isLoading: false,
        hits: [
            {
                "created_at": "2018-03-14T03:50:30.000Z",
                "title": "Stephen Hawking has died",
                "url": "http://www.bbc.com/news/uk-43396008",
                "author": "Cogito",
                "points": 6015,
                "story_text": null,
                "comment_text": null,
                "num_comments": 436,
                "story_id": null,
                "story_title": null,
                "story_url": null,
                "parent_id": null,
                "created_at_i": 1520999430,
                "relevancy_score": 8012,
                "_tags": ["story", "author_Cogito", "story_16582136"],
                "objectID": "16582136",
                "_highlightResult": {
                    "title": {
                        "value": "Stephen Hawking has died",
                        "matchLevel": "none",
                        "matchedWords": []
                    },
                    "url": {
                        "value": "http://www.bbc.com/news/uk-43396008",
                        "matchLevel": "none",
                        "matchedWords": []
                    },
                    "author": {
                        "value": "Cogito",
                        "matchLevel": "none",
                        "matchedWords": []
                    }
                }
            }
        ],
        hitsPerPage: 30,
        page: 0,
        nbPages: null,
        bookmarks: [],
        hidden: []
    }
})

describe('Renders Newstable Component', () => {
  const wrapper = mount(<Provider store={store}><NewsTable /></Provider>);

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should render table', () => {
    expect(wrapper.find("#news-table").hostNodes().length).toBe(1);
  });

  it('should render row', () => {
      console.log(wrapper.debug());
    expect(wrapper.find(".news-row").hostNodes().length).toBe(1);
  })

});
