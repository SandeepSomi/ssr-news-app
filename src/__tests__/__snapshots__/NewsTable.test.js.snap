// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Newstable Component should render correctly 1`] = `ReactWrapper {}`;

exports[`renders learn react link should render correctly 1`] = `ReactWrapper {}`;
